File: app/src/androidTest/java/com/carbit/inappkeyboard/KeyboardEngineInstrumentedTest.java
package com.carbit.inappkeyboard;

import android.text.InputType;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;

import androidx.test.core.app.ActivityScenario;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.carbit.inappkeyboard.keyboard.CandidateBarView;
import com.carbit.inappkeyboard.keyboard.InAppKeyboardPanelView;
import com.carbit.inappkeyboard.keyboard.InAppKeyboardView;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class KeyboardEngineInstrumentedTest {

    private void assertLayoutCycle(InAppKeyboardView keyboard, List<InAppKeyboardView.Layout> expected) {
        for (int i = 0; i < expected.size(); i++) {
            assertEquals("layout mismatch at step " + i, expected.get(i), keyboard.getCurrentLayout());
            if (i != expected.size() - 1) keyboard.injectKey("lang");
        }
    }

    @Test
    public void text_mode_injectKey_appends_text() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_text);

                et.setText("");
                panel.attachTo(et);

                keyboard.injectKey("a");
                keyboard.injectKey("b");
                keyboard.injectKey("c");

                assertEquals("abc", et.getText().toString());
            });
        }
    }

    @Test
    public void text_mode_space_enter_backspace() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_text);

                et.setText("");
                panel.attachTo(et);

                keyboard.injectKey("a");
                keyboard.injectKey("space");
                keyboard.injectKey("b");
                keyboard.injectKey("enter");
                keyboard.injectKey("c");
                keyboard.injectKey("⌫");

                assertEquals("a b\n", et.getText().toString());
            });
        }
    }

    @Test
    public void symbols_toggle_123_and_back_to_abc() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_text);

                et.setText("");
                panel.attachTo(et);

                keyboard.injectKey("123");
                keyboard.injectKey("@");
                keyboard.injectKey("abc");
                keyboard.injectKey("a");

                assertEquals("@a", et.getText().toString());
            });
        }
    }

    @Test
    public void inputMode_auto_infers_layouts_from_inputType() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();

                EditText etText = activity.findViewById(R.id.et_text);
                int textClass = etText.getInputType() & InputType.TYPE_MASK_CLASS;
                assertEquals(InputType.TYPE_CLASS_TEXT, textClass);

                // Ensure auto inference is enabled for this test.
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(etText);
                assertEquals(InAppKeyboardView.InputMode.TEXT, keyboard.getInputMode());

                EditText etNumber = activity.findViewById(R.id.et_number);
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(etNumber);
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.getCurrentLayout());

                EditText etPhone = activity.findViewById(R.id.et_phone);
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(etPhone);
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.getCurrentLayout());

                EditText etPassword = activity.findViewById(R.id.et_password);
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(etPassword);
                assertEquals(InAppKeyboardView.Layout.EN, keyboard.getCurrentLayout());
            });
        }
    }

    @Test
    public void text_mode_lang_cycles_EN_ZH_FR_AR_EN() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_text);

                panel.attachTo(et);
                keyboard.setInputMode(InAppKeyboardView.InputMode.TEXT);
                keyboard.setLayout(InAppKeyboardView.Layout.EN);

                assertLayoutCycle(keyboard, Arrays.asList(
                        InAppKeyboardView.Layout.EN,
                        InAppKeyboardView.Layout.ZH_PINYIN,
                        InAppKeyboardView.Layout.FR,
                        InAppKeyboardView.Layout.AR,
                        InAppKeyboardView.Layout.EN
                ));
            });
        }
    }

    @Test
    public void number_mode_symbols_toggle_returns_to_numeric() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_number);

                et.setText("");
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(et);
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.getCurrentLayout());

                keyboard.injectKey("123");
                assertEquals(InAppKeyboardView.Layout.SYMBOLS, keyboard.getCurrentLayout());

                keyboard.injectKey("1");
                keyboard.injectKey("abc");
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.getCurrentLayout());
                keyboard.injectKey("2");

                assertEquals("12", et.getText().toString());
            });
        }
    }

    @Test
    public void number_mode_injectKey_appends_digits() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                EditText et = activity.findViewById(R.id.et_number);

                et.setText("");
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(et);

                keyboard.injectKey("1");
                keyboard.injectKey("2");
                keyboard.injectKey("3");

                assertEquals("123", et.getText().toString());
            });
        }
    }

    @Test
    public void password_mode_lang_key_does_not_cycle_and_no_candidates() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                CandidateBarView candidateBar = panel.getCandidateBarView();
                EditText et = activity.findViewById(R.id.et_password);

                et.setText("");
                keyboard.setInputMode(InAppKeyboardView.InputMode.AUTO);
                panel.attachTo(et);

                InAppKeyboardView.Layout before = keyboard.getCurrentLayout();
                keyboard.injectKey("lang");
                InAppKeyboardView.Layout after = keyboard.getCurrentLayout();

                assertEquals(before, after);
                assertFalse(candidateBar.getVisibility() == View.VISIBLE);
            });
        }
    }

    @Test
    public void zh_pinyin_generates_candidates_and_click_commit() {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            scenario.onActivity(activity -> {
                InAppKeyboardPanelView panel = activity.findViewById(R.id.keyboard_panel);
                InAppKeyboardView keyboard = panel.getKeyboardView();
                CandidateBarView candidateBar = panel.getCandidateBarView();
                EditText et = activity.findViewById(R.id.et_text);

                et.setText("");
                panel.attachTo(et);
                keyboard.setLayout(InAppKeyboardView.Layout.ZH_PINYIN);

                keyboard.injectKey("n");
                keyboard.injectKey("i");

                LinearLayout container = candidateBar.findViewById(com.carbit.inappkeyboard.keyboard.R.id.candidates_container);
                assertTrue("candidate bar should be visible", candidateBar.getVisibility() == View.VISIBLE);
                assertTrue("expected >= 1 candidate", container.getChildCount() >= 1);

                container.getChildAt(0).performClick();
                String s = et.getText().toString();
                assertTrue("expected committed text after candidate click", s != null && !s.isEmpty());
            });
        }
    }
}

File: app/src/androidTest/java/com/carbit/inappkeyboard/KeyboardEngineInstrumentedTest.kt
package com.carbit.inappkeyboard

import android.text.InputType
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import androidx.test.core.app.ActivityScenario
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.carbit.inappkeyboard.keyboard.InAppKeyboardPanelView
import com.carbit.inappkeyboard.keyboard.InAppKeyboardView
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.assertFalse
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class KeyboardEngineInstrumentedTest {

    private fun assertLayoutCycle(keyboard: InAppKeyboardView, expected: List<InAppKeyboardView.Layout>) {
        expected.forEachIndexed { i, layout ->
            assertEquals("layout mismatch at step $i", layout, keyboard.currentLayout)
            if (i != expected.lastIndex) keyboard.injectKey("lang")
        }
    }

    @Test
    fun text_mode_injectKey_appends_text() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_text)

                et.setText("")
                panel.attachTo(et)

                keyboard.injectKey("a")
                keyboard.injectKey("b")
                keyboard.injectKey("c")

                assertEquals("abc", et.text.toString())
            }
        }
    }

    @Test
    fun text_mode_space_enter_backspace() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_text)

                et.setText("")
                panel.attachTo(et)

                keyboard.injectKey("a")
                keyboard.injectKey("space")
                keyboard.injectKey("b")
                keyboard.injectKey("enter")
                keyboard.injectKey("c")
                keyboard.injectKey("⌫")

                assertEquals("a b\n", et.text.toString())
            }
        }
    }

    @Test
    fun symbols_toggle_123_and_back_to_abc() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_text)

                et.setText("")
                panel.attachTo(et)

                // Enter symbols
                keyboard.injectKey("123")
                keyboard.injectKey("@")
                // Return to letters (symbols layout uses 'abc')
                keyboard.injectKey("abc")
                keyboard.injectKey("a")

                assertEquals("@a", et.text.toString())
            }
        }
    }

    @Test
    fun inputMode_auto_infers_layouts_from_inputType() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView

                // TEXT
                val etText = a.findViewById<EditText>(R.id.et_text)
                val textClass = etText.inputType and InputType.TYPE_MASK_CLASS
                assertEquals(InputType.TYPE_CLASS_TEXT, textClass)

                panel.attachTo(etText)
                assertEquals(InAppKeyboardView.InputMode.TEXT, keyboard.inputMode)

                // NUMBER
                val etNumber = a.findViewById<EditText>(R.id.et_number)
                panel.attachTo(etNumber)
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.currentLayout)

                // PHONE (treated as numeric)
                val etPhone = a.findViewById<EditText>(R.id.et_phone)
                panel.attachTo(etPhone)
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.currentLayout)

                // PASSWORD should land on EN.
                val etPassword = a.findViewById<EditText>(R.id.et_password)
                panel.attachTo(etPassword)
                assertEquals(InAppKeyboardView.Layout.EN, keyboard.currentLayout)
            }
        }
    }

    @Test
    fun text_mode_lang_cycles_EN_ZH_FR_AR_EN() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_text)

                panel.attachTo(et)
                keyboard.setLayout(InAppKeyboardView.Layout.EN)

                assertLayoutCycle(
                    keyboard,
                    listOf(
                        InAppKeyboardView.Layout.EN,
                        InAppKeyboardView.Layout.ZH_PINYIN,
                        InAppKeyboardView.Layout.FR,
                        InAppKeyboardView.Layout.AR,
                        InAppKeyboardView.Layout.EN,
                    ),
                )
            }
        }
    }

    @Test
    fun number_mode_symbols_toggle_returns_to_numeric() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_number)

                et.setText("")
                panel.attachTo(et)
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.currentLayout)

                // Toggle to symbols from numeric
                keyboard.injectKey("123")
                assertEquals(InAppKeyboardView.Layout.SYMBOLS, keyboard.currentLayout)

                // Note: number EditText will reject non-digit chars; use a digit that exists on symbols.
                keyboard.injectKey("1")

                // Back to numeric due to inputMode NUMBER
                keyboard.injectKey("abc")
                assertEquals(InAppKeyboardView.Layout.NUMERIC, keyboard.currentLayout)
                keyboard.injectKey("2")

                assertEquals("12", et.text.toString())
            }
        }
    }

    @Test
    fun number_mode_injectKey_appends_digits() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val et = a.findViewById<EditText>(R.id.et_number)

                et.setText("")
                panel.attachTo(et)

                keyboard.injectKey("1")
                keyboard.injectKey("2")
                keyboard.injectKey("3")

                assertEquals("123", et.text.toString())
            }
        }
    }

    @Test
    fun password_mode_lang_key_does_not_cycle_and_no_candidates() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val candidateBar = panel.candidateBarView
                val et = a.findViewById<EditText>(R.id.et_password)

                et.setText("")
                panel.attachTo(et)

                // Try to change language
                val before = keyboard.currentLayout
                keyboard.injectKey("lang")
                val after = keyboard.currentLayout

                // Should stick to EN in password mode
                assertEquals(before, after)

                // Candidates should remain hidden
                assertFalse(candidateBar.visibility == View.VISIBLE)
            }
        }
    }

    @Test
    fun zh_pinyin_generates_candidates_and_click_commit() {
        ActivityScenario.launch(MainActivity::class.java).use { scenario ->
            scenario.onActivity { a ->
                val panel = a.findViewById<InAppKeyboardPanelView>(R.id.keyboard_panel)
                val keyboard = panel.keyboardView
                val candidateBar = panel.candidateBarView
                val et = a.findViewById<EditText>(R.id.et_text)

                et.setText("")
                panel.attachTo(et)
                keyboard.setLayout(InAppKeyboardView.Layout.ZH_PINYIN)

                keyboard.injectKey("n")
                keyboard.injectKey("i")

                val container = candidateBar.findViewById<LinearLayout>(
                    com.carbit.inappkeyboard.keyboard.R.id.candidates_container
                )
                assertTrue("candidate bar should be visible", candidateBar.visibility == View.VISIBLE)
                assertTrue("expected >= 1 candidate", container.childCount >= 1)

                // Click first candidate should commit Chinese text into EditText.
                container.getChildAt(0).performClick()
                val s = et.text.toString()
                assertTrue("expected committed text after candidate click", s.isNotEmpty())
            }
        }
    }
}
